databuilder/query_model.py

```diff
@@ -365,38 +359,83 @@ class DomainMismatchError(Exception):
+    #    as long as any filters applied to the data can be arranged in a linear
+    #    sequence. That is, it's legal to combine `Table A -> Filter B -> Filter C` and
+    #    `Table A -> Filter B`. But it's not legal to combine `Table A -> Filter D` and
+    #    `Table A -> Filter E`. There's never a need to use such constructions and they
+    #    will almost always be the result of user error. It's better to reject them
+    #    immediately than give the user unexpected results at query time.
+    #
+    # We enforce this by modelling domains as a hierarchy. At the root is the patient
+    # domain, each table has an unique domain descending from this, and each filter
+    # operation creates a new domain descending from the domain of its source. A valid
+    # operation is one where the domains of its inputs are "nested" in the sense that
+    # each domain is an ancestor or descendant of every other.
+    #
+    # We use sets to implement this hierarchy: we start with the patient domain as a set
+    # containing a single unique value, and we create descendent domains by forming the
+    # union of the parent domain with a new unqiue value. The `issubset()` relation then
+    # gives us the "is domain ancestor" semantics we want.
+    #
+    # Another way of expressing the "nested" condition above is that the input domains
+    # must be totally ordered by ancestory. We can check this by sorting them (sets
+    # naturally sort by subset-hood in Python) and then confirming that the ordering is
+    # total by testing that each adjacent pair has the required subset relation.
+    domains = sorted(get_input_domains(node))
+    for a, b in zip(domains, domains[1:]):
+        if not a.issubset(b):
+            raise DomainMismatchError(
+                f"Attempt to combine unrelated domains:\n{a} and {b}"
+                f"\nIn node:\n{node}"
+            )
+
 
+def get_input_domains(node):
+    return {get_domain(input_node) for input_node in get_input_nodes(node)}
 
-def validate_inputs_have_common_domain(node):
-    domains = get_domains(node)
-    non_patient_domains = domains - {PATIENT_DOMAIN}
-    if len(non_patient_domains) > 1:
-        raise DomainMismatchError(
-            f"Attempt to combine multiple domains:\n{non_patient_domains}"
-            f"\nIn node:\n{node}"
-        )
 
+# We use an arbitrary unique object to represent the patient domain
+PATIENT_DOMAIN = frozenset([object()])
 
-# For most operations, their domain is the just the domains of all their inputs
+
+# For most operations, their domain is the "smallest" of the domains of their inputs
+# i.e. the one furthest from the root
 @singledispatch
-def get_domains(node):
-    return set().union(
-        *[get_domains(input_node) for input_node in get_input_nodes(node)]
-    )
+def get_domain(node):
+    return sorted(get_input_domains(node))[-1]
+
+
+# Selecting a table creates a new domain which descends from the patient domain
```

Comment:
Specifically selecting a non-patient table? Maybe "table" (as in `SelectTable`) isn't a specific enough term and we need something that's more clearly in opposition to "patient table".